alibi/explainers/similarity/grad.py

```diff
@@ -0,0 +1,249 @@
+        """Explain the predictor's predictions for a given input.
+
+        Computes the similarity score between the input and the training set. Returns an explainer object containing
+        the scores, the indices of the training set instances sorted by descending similarity to the input and the
+        most similar and least similar instances of the data set.
+
+        Parameters
+        ----------
+        X
+            `X` can be a `numpy` array, `tensorflow` tensor, or `pytorch` tensor of the same shape as the training data
+            with or without a leading batch dimension. If the batch dimension is missing it's added.
+        Y
+            `Y` can be a `numpy` array, `tensorflow` tensor or a `pytorch` tensor. In the case of a regression task, the
+            `Y` argument must be present. If the task is classification then `Y` defaults to the model prediction.
+
+        Returns
+        -------
+        `Explanation` object containing the ordered similarity scores for the instance with additional metadata as \
+        attributes. Contains the following data-related attributes
+            -  `scores`: ``np.ndarray`` - similarity scores for each instance in the training set sorted in descending \
+            order.
+            -  `ordered_indices`: ``np.ndarray`` - indices of the training set instances sorted by the similarity \
+            score in descending order.
+            -  `most_similar`: ``np.ndarray`` - most similar instances to the input.
+            -  `least_similar`: ``np.ndarray`` - least similar instances to the input.
+
+        Raises
+        -------
+        ValueError
+            If `Y` is ``None`` and the `task` is ``'regression'``.
+        ValueError
+            If the shape of `X` or `Y` does not match the shape of the training or target data
+        ValueError
+            If the fit method has not been called prior to calling this method.
+        """
+        self._verify_fit()
+        X, Y = self._preprocess_args(X, Y)
+        grad_X_test = self._compute_grad(X, Y)
+        if not self.precompute_grads:
+            scores = self._compute_adhoc_similarity(grad_X_test)
+        else:
+            scores = self.sim_fn(self.grad_X_train, grad_X_test)
+        return self._build_explanation(scores)
+
+    def _build_explanation(self, scores: np.ndarray) -> "Explanation":
+        """Builds an explanation object.
+
+        Parameters
+        ----------
+        scores
+            The scores for each of the instances in the data set computed by the similarity method.
+        """
+        data = copy.deepcopy(DEFAULT_DATA_SIM)
+        sorted_score_indices = np.argsort(scores)[::-1]
+        data.update(
+            scores=scores[sorted_score_indices],
+            ordered_indices=sorted_score_indices,
+            most_similar=self.X_train[sorted_score_indices[0]],
+            least_similar=self.X_train[sorted_score_indices[-1]],
```

Comment:
not sure if maybe we should go for top 5 ...

Reply:
Hmm I didn't actually notice this return. I foresee that with Seldon Deploy it would not be enough to return the indices as the training set to index into wouldn't be easily available so we will have to return the actual instances (and, as a result, probably make the number configurable as an argument to `explain`).

In the meantime I think we can go with Robert's suggestion and hard-code returning top 5 (update docstring too). Do we need to also return the least similar ones, what's the use case?